utils/client/client.c

```diff
@@ -0,0 +1,687 @@
+}
+
+static bool monitor_loop_on = true;
+static void monitor_loop_stop(int sig) {
+	/* Call to this handler restores the default action, so on the
+	 * second call the program will be forcefully terminated. */
+
+	struct sigaction sigact = { .sa_handler = SIG_DFL };
+	sigaction(sig, &sigact, NULL);
+
+	monitor_loop_on = false;
+}
+
+static int cmd_monitor(int argc, char *argv[]) {
+	if (argc != 1) {
+		print_error_usage("Invalid arguments.");
+		return EXIT_FAILURE;
+	}
+
+	bluealsa_dbus_connection_signal_match_add(&dbus_ctx,
+			service_name, NULL, BLUEALSA_INTERFACE_MANAGER, "PCMAdded", NULL);
+	bluealsa_dbus_connection_signal_match_add(&dbus_ctx,
+			service_name, NULL, BLUEALSA_INTERFACE_MANAGER, "PCMRemoved", NULL);
+
+	if (!dbus_connection_add_filter(dbus_ctx.conn, dbus_signal_handler, NULL, NULL)) {
+		print_error("Couldn't add D-Bus filter");
+		return EXIT_FAILURE;
+	}
+
+	struct sigaction sigact = { .sa_handler = monitor_loop_stop };
+	sigaction(SIGTERM, &sigact, NULL);
+	sigaction(SIGINT, &sigact, NULL);
+
+	while (monitor_loop_on) {
+
+		struct pollfd pfds[10];
+		nfds_t pfds_len = ARRAYSIZE(pfds);
+
+		if (!bluealsa_dbus_connection_poll_fds(&dbus_ctx, pfds, &pfds_len)) {
+			print_error("Couldn't get D-Bus connection file descriptors");
+			return EXIT_FAILURE;
+		}
+
+		if (poll(pfds, pfds_len, -1) == -1 &&
+				errno == EINTR)
+			continue;
+
+		if (bluealsa_dbus_connection_poll_dispatch(&dbus_ctx, pfds, pfds_len))
+			while (dbus_connection_dispatch(dbus_ctx.conn) == DBUS_DISPATCH_DATA_REMAINS)
+				continue;
+	}
+
+	return EXIT_SUCCESS;
+}
+
+static struct command {
+	const char *name;
+	int (*func)(int argc, char *arg[]);
+	const char *help;
+} commands[] = {
+	{ "get-codecs", cmd_get_codecs, "PCM_PATH", },
+	{ "list-pcms", cmd_list_pcms, "" },
+	{ "monitor", cmd_monitor, "" },
+	{ "mute", cmd_mute, "PCM_PATH y|n [y|n]" },
+	{ "open", cmd_open, "PCM_PATH" },
+	{ "properties", cmd_properties, "PCM_PATH" },
+	{ "select-codec", cmd_select_codec, "PCM_PATH CODEC" },
+	{ "set-volume", cmd_set_volume, "PCM_PATH N [N]" },
+	{ "softvol", cmd_softvol, "PCM_PATH y|n" },
+};
+
+static void usage(void) {
+	int index;
+	for (index = 0; index < ARRAYSIZE(commands); index++) {
+		printf("%s [options] %s %s\n", progname, commands[index].name, commands[index].help);
+	}
+	printf("options:\n");
+	printf("   -h, --help       Show this help\n");
+	printf("   -V, --version    Show version\n");
+	printf("   -B, --dbus=NAME  BlueALSA service name suffix\n");
+	printf("   -q, --quiet      Do not print any error messages\n");
+}
+
+int main(int argc, char *argv[]) {
+
+	const struct option long_options[] = {
+		{"dbus",        required_argument, NULL, 'B'},
+		{"version",     no_argument, NULL, 'V'},
+		{"help",        no_argument, NULL, 'h'},
+		{"quiet",       no_argument, NULL, 'q'},
+		{ 0 },
+	};
+
+	progname = argv[0];
+
+	int c;
+    while ((c = getopt_long(argc, argv, "+s:Vh", long_options, NULL)) != -1) {
```

Comment:
So now the `long_options` array and the `getopt_long()` are close enough to see that the `optstring` in the `getopt_long()` does not match with the array.

Reply:
Ooops ! My apologies, that's just laziness on my part. I haven't tested short options since a very early draft of this code.

Reply:
see commit 1d92201